apps/prairielearn/src/sync/syncFromDisk.ts

```diff
@@ -37,35 +34,56 @@ export async function syncDiskToSqlWithLock(
   courseDir: string,
   logger: Logger,
 ): Promise<SyncResults> {
+  async function timed<T>(label: string, fn: () => Promise<T>): Promise<T> {
+    const start = performance.now();
+
+    const result = await fn();
+
+    const duration = performance.now() - start;
+    logger.verbose(`${label} in ${duration.toFixed(2)}ms`);
+
+    return result;
+  }
+
   logger.info('Loading info.json files from course repository');
-  perf.start('sync');
 
-  const courseData = await perf.timed('loadCourseData', () =>
+  const courseData = await timed('Loaded course data from disk', () =>
     courseDB.loadFullCourse(courseId, courseDir),
   );
+
   logger.info('Syncing info to database');
-  await perf.timed('syncCourseInfo', () => syncCourseInfo.sync(courseData, courseId));
-  const courseInstanceIds = await perf.timed('syncCourseInstances', () =>
-    syncCourseInstances.sync(courseId, courseData),
-  );
-  await perf.timed('syncTopics', () => syncTopics.sync(courseId, courseData));
-  const questionIds = await perf.timed('syncQuestions', () =>
-    syncQuestions.sync(courseId, courseData),
-  );
 
-  await perf.timed('syncTags', () => syncTags.sync(courseId, courseData, questionIds));
-  await perf.timed('syncAssessmentSets', () => syncAssessmentSets.sync(courseId, courseData));
-  await perf.timed('syncAssessmentModules', () => syncAssessmentModules.sync(courseId, courseData));
-  perf.start('syncAssessments');
-  await Promise.all(
-    Object.entries(courseData.courseInstances).map(async ([ciid, courseInstanceData]) => {
-      const courseInstanceId = courseInstanceIds[ciid];
-      await perf.timed(`syncAssessments${ciid}`, () =>
-        syncAssessments.sync(courseId, courseInstanceId, courseInstanceData, questionIds),
+  await timed('Synced all course data', async () => {
+    await timed('Synced course info', () => syncCourseInfo.sync(courseData, courseId));
+    const courseInstanceIds = await timed('Synced course instances', () =>
+      syncCourseInstances.sync(courseId, courseData),
+    );
+    await timed('Synced topics', () => syncTopics.sync(courseId, courseData));
+    const questionIds = await timed('Synced questions', () =>
+      syncQuestions.sync(courseId, courseData),
+    );
+
+    await timed('Synced tags', () => syncTags.sync(courseId, courseData, questionIds));
+    await timed('Synced assessment sets', () => syncAssessmentSets.sync(courseId, courseData));
+    await timed('Synced assessment modules', () =>
+      syncAssessmentModules.sync(courseId, courseData),
+    );
+    await timed('Synced all assessments', async () => {
+      // Ensure that a single course with a ton of course instances can't
+      // monopolize the database connection pool.
```

Comment:
While testing all this, I realized that we would try to sync all course instances in parallel. Every call would queue up a DB query, so for courses with a lot of course instances (we have some with almost 20 in practice), a single sync would tie up the DB connection pool for many seconds. I think this explains the periodic "freezes" I've seen on instructor server routes.

Limiting concurrency has two benefits: it leaves other clients in the pool, and it also allows other queries to join the queue, making everything more fair.

Reply:
Great catch, we never want unlimited concurrency, I think.